src/main/JayZ.ts

```diff
@@ -1,49 +1,123 @@
+import { EncryptedItemMetadataV1, EncryptedJayZItem, LegacyEncryptedItemMetadata, MetadataVersion } from "./types"
 
-export interface JayZProps {
-  encryptor: Encryptor
+export type JayZConfig = {
+  keyProvider: DataKeyProvider
+  encryptor?: Encryptor
+  maxUsesPerDataKey?: number
 }
 
-export interface EncryptItemProps<T, U extends keyof T> {
+export type EncryptItemRequest<T, K extends keyof T> = {
   item: T
-  fieldsToEncrypt: U[]
+  fieldsToEncrypt: K[]
 }
 
 export class JayZ {
-  private encryptor: Encryptor
+  public readonly ready = ready
+  private keyProvider: DataKeyProvider
+  private encryptor: Encryptor = new LibsodiumEncryptor()
+  private maxUsesPerDataKey: number
+  private currentDataKey?: Promise<DataKey>
+  private currentDataKeyUsesRemaining: number
 
-  constructor(props: JayZProps) {
-    this.encryptor = props.encryptor
+  constructor(config: JayZConfig) {
+    this.keyProvider = config.keyProvider
+    this.encryptor =
+      config.encryptor !== undefined
+        ? config.encryptor
+        : new LibsodiumEncryptor()
+    this.maxUsesPerDataKey = config.maxUsesPerDataKey || 1
+    this.currentDataKeyUsesRemaining = this.maxUsesPerDataKey
   }
 
-  async encryptItem<T, U extends keyof T>(itemToEncrypt: EncryptItemProps<T, U>): Promise<EncryptedJayZItem<T, U>> {
+  async encryptItem<T, K extends keyof T>(
+    itemToEncrypt: EncryptItemRequest<T, K>
+  ): Promise<EncryptedJayZItem<T, K>> {
     const { item, fieldsToEncrypt } = itemToEncrypt
-    const { encryptedItem } = await this.encryptor.encrypt({ item, fieldsToEncrypt })
-    return encryptedItem
+    const { dataKey, encryptedDataKey } = await this.getNextDataKey()
+    const { plaintextFields, nonce, encryptedFields } = this.encryptor.encrypt({
+      item,
+      fieldsToEncrypt,
+      dataKey
+    })
+
+    const __jayz__metadata: EncryptedItemMetadataV1 = {
+      encryptedDataKey,
+      nonce,
+      scheme: this.encryptor.scheme,
+      encryptedFields,
+      metadataVersion: MetadataVersion.V1
+    }
+
+    return {
+      ...plaintextFields,
+      __jayz__metadata
+    }
   }
 
-  async encryptItems<T, U extends keyof T>(
-    itemsToEncrypt: EncryptItemProps<T, U>[]
-  ): Promise<EncryptedJayZItem<T, U>[]> {
+  encryptItems<T, K extends keyof T>(
+    itemsToEncrypt: EncryptItemRequest<T, K>[]
+  ): Promise<EncryptedJayZItem<T, K>[]> {
     if (itemsToEncrypt.length === 0) {
-      return []
+      return Promise.resolve([])
     }
 
     const items = itemsToEncrypt.map((item) => this.encryptItem(item))
     return Promise.all(items)
   }
 
-  async decryptItem<T, U extends keyof T>(item: EncryptedJayZItem<T, U>): Promise<T> {
-    const { decryptedItem } = await this.encryptor.decrypt<T, U>({ item })
+  async decryptItem<T, K extends keyof T>(
+    itemToDecrypt: EncryptedJayZItem<T, K>
+  ): Promise<T> {
+    const {
+      nonce,
+      encryptedDataKey,
+    } = itemToDecrypt.__jayz__metadata
```

Comment:
At first glance, I don't follow how this will be backward compatible with item encryption that's been done by using the previous approach that kept the encrypted fields at the same level, rather than putting them within `__jayz__metadata`.

Maybe we don't care about backward compatibility since Vault is already using the version that we're checking in with this PR? Do we need to consider other folks using Jay-Z since it's a public repo?

Reply:
@gbryan 

> At first glance, I don't follow how this will be backward compatible with item encryption that's been done by using the previous approach that kept the encrypted fields at the same level, rather than putting them within __jayz__metadata.

Actually, both "legacy" _and_ "v1" use the same metadata field name. The primary difference is: 
 
1. "legacy" puts the a set of field names to decrypt into the metadata field. 
2. "v1" puts the the encrypted field values into the metadata field. 

In these changes the `Encryptor` interface was updated to handle both via its signature: `decrypt<...>(params: DecryptParams<...> | LegacyDecryptParams<...>): DecryptResult<...>`. 

Since the item to be decrypted here can either be legacy or v1, lines 80 and 81 force cast to each type. And then the call to `.decrypt` on line 87 ends up handling both cases as the parameters being passed in will either match the "legacy" signature or the "v1" signature given either `fieldsToDecrypt` or `encryptedFields` will be empty. 

This reads as me trying to be too clever/cute, it'd probably be clearer to readers if there was just a slightly longer, but more explicit check on which type of item we're dealing with.

Reply:
> This reads as me trying to be too clever/cute, it'd probably be clearer to readers if there was just a slightly longer, but more explicit check on which type of item we're dealing with.

I changed the logic so that we're not force casting the type: https://github.com/HeadspaceMeditation/jay-z/pull/39/commits/16fddff049024c0d03671574e0ee4b7f7f060dfb

Reply:
@gbryan great catch! I added a check to see if `encryptedFields` is null / undefined.

Reply:
Great! If I'm reading it right, I don't think the `!('encryptedFields' in params)` will ever be true, so I think we _do_ need the `undefined` check that you added.

It'd be nice to remove that `!('encryptedFields' in params)` check as well (I know from chatting on Slack that you're already looking into it!).